### ISSUE TITLE
GENIE3

### ISSUE THREAD

**Issue by s-aibar** [Bioconductor/Contributions#321]
- Repository: https://github.com/aertslab/GENIE3

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by mtmorgan** [comment]
@vobencha will review this package, but I note it has no Bioconductor dependencies. A key reason to contribute to Bioconductor is to facilitate interaction with other packages in the Bioconductor ecosystem through shared data structures. It would be natural to use SummarizedExperiment or Biobase::ExpressionSet for representing expression input data. One might also anticipate GSEABase for representing gene sets and / or the graph package for graph representation (I did not look at your package in detail so do not know how relevant these specific representations are.

**Comment by vobencha** [comment]
Before I do a review, please incorporate comments from https://github.com/Bioconductor/Contributions/issues/321#issuecomment-286548000 or explain why you think they don't apply to GENIE3.

Valerie

**Comment by vobencha** [comment]
Any updates? Should we close this issue?
Valerie

**Comment by s-aibar** [comment]
Hello,

Sorry for not replying earlier, we are creating this package as part of a workflow. Some of the other packages submitted at the same time also require some changes to include in Bioconductor, so I was considering whether some further changes need to be done before you review this one.

In principle, I think the previous comment does not apply to this package. GENIE3 package does not have any Bioconductor dependencies because its only input is an expression matrix (i.e. `exprs()` from an ExpressionSet, it doesnt need use the rest), and a list of gene identifiers (input as a character vector, matching the row names of the expression matrix). If you consider that it should accept a specific class as input, I can certainly modify the package, but a priory I think it would only make it more complex (e.g. there are multiple "containers" for expression matrices, SummarizedExperiment, ExpressionSet, SCESet, etc... are them all compatible?). 

Regarding the networks: The package is meant to reconstruct gene-regulatory networks, but the network is returned as an incidence matrix or an edge list (data.frame containing the columns 'TF' and 'target'). We kept these formats for compatibility with previous versions implemented in Python and Matlab, and because it is easy to export/convert into any other format (e.g. export as text file, or convert into R graph objects with igraph::graph_from_data_frame()). We can consider exporting it to other formats if you feel it is appropriate, but at the moment they are not used within the package.

Thank you for taking the package into consideration,
Sara Aibar

**Comment by vobencha** [comment]
The Bioconductor classes mentioned (SummarizeExperiment, ExpressionSet, etc.) all have a slot that contains a matrix of data; the classes are not identical but the matrix object is. The idea is that you'd write a generic with methods for the different classes - each method would extract the matrix from the Bioconductor class using the appropriate accessor function. Once the matrix is extracted the rest of the code would be the same for all methods.

You mentioned you're working on other aspects of a pipeline - maybe the functionality of GENIE3 should be folded into one of those - I don't think it requires a stand alone package. If you agree we can close this issue.

Valerie

**Comment by s-aibar** [comment]
Ok, we can easily accept those objects as input to GENIE3, and extract the matrix already within the function (although personally I find it safer if it is the user who decides which slot to use...).  

Regarding the independence of GENIE3: we are using GENIE3 as part of an analysis pipeline, including AUCell (which was also submitted as a new package), and motif and GO enrichment. However, the other packages don't really have much to do with co-expression networks. For example, AUCell's aim is to evaluate the activity of gene signatures on single-cell RNA-seq... If needed, it could be merged with some of these packages, but conceptually they are independent tools, they do different type of analyses and they can be applied on different types of data. In this way, it would be better to keep them as independent packages (for maintenance, reduce dependence across packages, easier to understand for the user...). Furthermore, GENIE3 is already known as an independent tool (winner of the DREAM4 and 5 challenges, also available in python and matlab at http://www.montefiore.ulg.ac.be/~huynh-thu/GENIE3.html), and it will be easier to find if the new R implementation can be kept as a package with its own name. Its current [citation stats](http://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0012776#citedHeader) also support the idea that it is useful as an independent package by itself. Whether it is worth to include it in Bioconductor or not, of course it is up to you to decide, but since GENIE3 aim is to infer gene regulatory networks, I certainly think it fits into the description of "tools for the analysis and comprehension of high-throughput genomic data", and therefore it fits better in Bioconductor than other repositories such as R-cran which are more generic...

Regards, 
Sara Aibar

**Comment by vobencha** [comment]
Hi Sara,

It's fine to keep GENIE3 as it's own package. I suggested breaking it up because I thought you were leaning that way - I must have misunderstood. 

There isn't much point in having a package in Bioconductor that doesn't use the infrastructure (could just submit to CRAN) so you do need to support the core classes mentioned in these comments:
https://github.com/Bioconductor/Contributions/issues/321#issuecomment-286548000
https://github.com/Bioconductor/Contributions/issues/321#issuecomment-297994406

You mention that you 'find it safer to let the user decided which slot to use'. The point of S4 generics and method dispatch is to provide a level of convenience to the user. For example, the user has an object that holds discrete data and wants to invoke a method for counting - instead of having to figure out where the data are, get the name of the extractor, extract the data, feed it to the method etc. they simply invoke the method on the class. There aren't multiple slots in an object (at least not the objects we suggested) that have the same type of data; the idea that the user would 'choose the slot' doesn't make sense because only one meets the criteria.

In https://github.com/Bioconductor/Contributions/issues/321#issuecomment-294945030 you said 
```
"... GENIE3 package does not have any Bioconductor dependencies because its only input is an expression matrix (i.e. exprs() from an ExpressionSet, it doesnt need use the rest), and a list of gene identifiers (input as a character vector, matching the row names of the expression matrix). 
..."
```
The ExpressionSet object (as well as the others) have slots that contain the gene identifiers so in fact you could use 2 slots from the object. Your method could extract the matrix of data and the identifiers and pass them on to the main body of your function which would be reused by multiple methods. Does that make sense? I can explain more (with code example) if necessary.

Valerie

**Comment by vobencha** [comment]
Are you planning to submit a new version?
Valerie

**Comment by s-aibar** [comment]
Dear Valerie, 

I have just submitted a new version, which accepts an ExpressionSet or SummarizedExperiment as input. 

However, I am not sure if the way I have used to extract the expression matrix is the best, as it does not make use of class inheritance, etc... Let me know if there is any standard/better approach:
`
    # Biobase::ExpressionSet
    if("ExpressionSet" %in% class(exprMatrix)) 
    {
        exprMatrix <- Biobase::exprs(exprMatrix)
    }
    
    # SummarizedExperiment::RangedSummarizedExperiment
    # Use SummarizedExperiment rather than the older ExpressionSet, especially for sequence data.
    if(grepl("SummarizedExperiment", class(exprMatrix))) 
    {
        usedSlot <- names(SummarizedExperiment::assays(exprMatrix))
        otherAvailable <- names(SummarizedExperiment::assays(exprMatrix))[-1]
        
        exprMatrix <- SummarizedExperiment::assay(exprMatrix)
        
        msg <- paste0("SummarizedExperiment slot extracted as expression matrix: \t", usedSlot) 
        if(length(otherAvailable)>0) msg <- paste(msg, "\nOther slots available:", paste(otherAvailable, collapse=", "))
        warning(msg)
    }
`

Also, in the new version, the Windows building is giving an error (related to the C function). Do you know what can be wrong? 

`
make: C:/Users/BIOCBU~1/BBS-3~1.6-B/R/etc/x64/Makeconf: Permission denied
gcc    -c -o GENIE3.o GENIE3.c
GENIE3.c:19:15: fatal error: R.h: No such file or directory
`

Thank you,
Sara

**Comment by vobencha** [comment]
Thanks for the update. I'll have a look at this tomorrow.
Valerie

**Comment by vobencha** [comment]
Hi,
The windows error was a problem on our end. It should be cleaned up now.

Examples of writing S4 methods are in IRanges, GenomicRanges, GenomicFeatures, Rsamtools, Biobase and other core packages. Here is an example of how you would convert GENE3() to S4:
```
setGeneric("GENE3", signature="exprMatrix",
    function(exprMatrix, treeMethod="RF", K="sqrt", nTrees=1000, regulators=NULL, targets=NULL, nCores=1, verbose=FALSE, seed=NULL)
{
    standardGeneric("GENE3")
}

setMethod("GENE3", "SummarizedExperiment",
    function(exprMatrix, treeMethod="RF", K="sqrt", nTrees=1000, regulators=NULL, targets=NULL, nCores=1, verbose=FALSE, seed=NULL)
{
    .GENE3 (assays(exprMatrix), treeMethod, K, nTrees, regulators, targets, nCores, verbose, seed)
}

setMethod("GENE3", "ExpressionSet",
    function(exprMatrix, treeMethod="RF", K="sqrt", nTrees=1000, regulators=NULL, targets=NULL, nCores=1, verbose=FALSE, seed=NULL)
{
    .GENE3 (exprs(exprMatrix), treeMethod, K, nTrees, regulators, targets, nCores, verbose, seed)
}
```
This internal helper would do everything that GENE3() currently does starting ~ line 79 with the arg checking:
```
.GENE3 <- function(exprMatrix, treeMethod="RF", K="sqrt", nTrees=1000, regulators=NULL, targets=NULL, nCores=1, verbose=FALSE, seed=NULL)
{
    # check input arguments
    if (!is.matrix(exprMatrix) && !is.array(exprMatrix)) {
		stop("Parameter exprMatrix must be a two-dimensional matrix where 
                        each row corresponds to a gene and each column corresponds 
                        to a condition/sample.")
    }
    ...
}
```
Valerie

**Comment by vobencha** [comment]
Hi Sara,
We need to finish up this review. Do you have questions about converting GENE3() to a generic with methods?
Valerie

**Comment by s-aibar** [comment]
Hello Valerie,

GENIE3() is now updated as a Generic.
Sorry, I didn't understand what you meant about the generics in first place, but with the examples it was very easy. Thank you!

**Comment by vobencha** [comment]
Great. Please bump the version so we get a new build.
Valerie

**Comment by vobencha** [comment]
Hi,
Just a few last things to clean up in the SPB output:
```
* checking dependencies in R code ... NOTE
'library' or 'require' calls in package code:
  doParallel doRNG foreach
  Please use :: or requireNamespace() instead.
  See section 'Suggested packages' in the 'Writing R Extensions' manual.
Namespace in Imports field not imported from: testthat
  All declared Imports should be used.
package 'methods' is used but not declared

* checking R code for possible problems ... NOTE
.GENIE3: no visible global function definition for %dorng%
Undefined global functions or variables:
  %dorng%

* Checking DESCRIPTION/NAMESPACE consistency...
    * WARNING: Import testthat in NAMESPACE as well as DESCRIPTION.
```
Valerie

**Comment by vobencha** [comment]
Thanks. Looks good.
